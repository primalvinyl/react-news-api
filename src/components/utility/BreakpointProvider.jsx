import React from 'react';

const BreakpointContext = React.createContext({
    'xs': false,
    'smUp': false,
    'mdUp': false,
    'lgUp': false,
    'xlUp': false,
    'smDown': false,
    'mdDown': false,
    'lgDown': false,
    'xlDown': false
});

export const BreakpointProvider = ({ children }) => {
    const [queryMatch, setQueryMatch] = React.useState({});

    const getMediaQuery = React.useCallback(() => {
        const queries = {
            'xs': '(max-width: 599px)',
            'smUp': '(min-width: 600px)',
            'mdUp': '(min-width: 900px)',
            'lgUp': '(min-width: 1200px)',
            'xlUp': '(min-width: 1800px)',
            'smDown': '(max-width: 600px)',
            'mdDown': '(max-width: 900px)',
            'lgDown': '(max-width: 1200px)',
            'xlDown': '(max-width: 1800px)'
        };
        const mediaQueryLists = {};
        const keys = Object.keys(queries);
        let isAttached = false;

        const handleQueryListener = () => {
            const updatedMatches = keys.reduce((acc, media) => {
                acc[media] = !!(mediaQueryLists[media] && mediaQueryLists[media].matches);
                return acc;
            }, {})
            setQueryMatch(updatedMatches);
        };

        if (window && window.matchMedia) {
            const matches = {};
            keys.forEach(media => {
                if (typeof queries[media] === 'string') {
                    mediaQueryLists[media] = window.matchMedia(queries[media]);
                    matches[media] = mediaQueryLists[media].matches;
                } else {
                    matches[media] = false;
                }
            });
            setQueryMatch(matches);
            isAttached = true;
            keys.forEach(media => {
                if (typeof queries[media] === 'string') {
                    mediaQueryLists[media].addListener(handleQueryListener);
                }
            });
        }

        return () => {
            if (isAttached) {
                keys.forEach(media => {
                    if (typeof queries[media] === 'string') {
                        mediaQueryLists[media].removeListener(handleQueryListener);
                    }
                });
            }
        };
    }, []);

    React.useEffect(() => getMediaQuery(), [getMediaQuery]);

    return (
        <BreakpointContext.Provider value={ queryMatch }>
            {children}
        </BreakpointContext.Provider>
    )
}

export function useBreakpoint() {
    const context = React.useContext(BreakpointContext);
    return context;
}
