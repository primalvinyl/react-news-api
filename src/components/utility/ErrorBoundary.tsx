import React from 'react';
import ErrorPage from '../../routes/ErrorPage';

type ErrorBoundaryProps = {
    readonly children: JSX.Element;
};

type ErrorBoundaryState = {
    hasError: boolean;
};

class ErrorBoundary extends React.Component<ErrorBoundaryProps, ErrorBoundaryState> {
    constructor(props: ErrorBoundaryProps) {
        super(props);
        this.state = { hasError: false };
    }
    static getDerivedStateFromError() {
        return { hasError: true };
    }
    render(): React.ReactElement {
        if (this.state.hasError) {
            return <ErrorPage />
        }
        else return this.props.children;
    }
}

export default ErrorBoundary;
