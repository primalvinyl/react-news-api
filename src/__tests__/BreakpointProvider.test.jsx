import React from 'react';
import { shallow } from 'enzyme';
import Renderer from 'react-test-renderer';
import { BreakpointProvider } from '../components/utility/BreakpointProvider';

describe('BreakpointProvider', () => {
    it('renders', () => {
        const myWrapper = shallow(<BreakpointProvider />);
        expect(myWrapper.exists());
    });
    it('matches snapshot', () => {
        const component = Renderer.create(<BreakpointProvider />);
        expect(component.toJSON()).toMatchSnapshot();
    });
});
