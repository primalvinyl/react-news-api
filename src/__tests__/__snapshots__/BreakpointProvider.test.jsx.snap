// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreakpointProvider matches snapshot 1`] = `null`;
